charmcraft/charm_builder.py

```diff
@@ -0,0 +1,357 @@
+        # in the directory is not linking directly to the entrypoint, and also check all the
+        # mandatory ones are present
+        dest_hookpath = self.buildpath / HOOKS_DIR
+        if not dest_hookpath.exists():
+            dest_hookpath.mkdir()
+
+        # get those built hooks that we need to replace because they are pointing to the
+        # entrypoint directly and we need to fix the environment in the middle
+        current_hooks_to_replace = []
+        for node in dest_hookpath.iterdir():
+            if node.resolve() == linked_entrypoint:
+                current_hooks_to_replace.append(node)
+                node.unlink()
+                logger.debug(
+                    "Replacing existing hook %r as it's a symlink to the entrypoint",
+                    node.name,
+                )
+
+        # include the mandatory ones and those we need to replace
+        hooknames = MANDATORY_HOOK_NAMES | {x.name for x in current_hooks_to_replace}
+        for hookname in hooknames:
+            logger.debug("Creating the %r hook script pointing to dispatch", hookname)
+            dest_hook = dest_hookpath / hookname
+            if not dest_hook.exists():
+                relative_link = relativise(dest_hook, dispatch_path)
+                dest_hook.symlink_to(relative_link)
+
+    def handle_dependencies(self):
+        """Handle from-directory and virtualenv dependencies."""
+        logger.debug("Installing dependencies")
+
+        if not self.python_packages and not self.requirements:
+            return
+
+        # create venv so packages can be cached. note: additional packages
+        # pip and setuptools will also be installed!
+        staging_venv_dir = self.charmdir / STAGING_VENV_DIRNAME
+        logger.debug("create staging venv in %r", str(staging_venv_dir))
+        retcode = _process_run([sys.executable, "-m", "venv", str(staging_venv_dir)])
+        if retcode:
+            raise CommandError("problems creating venv")
+
+        retcode = _process_run(["pip3", "--version"])
+        if retcode:
+            raise CommandError("problems using pip")
+
+        cmd = ["pip3", "install"]
+
+        if not self.allow_pip_binary:
+            cmd.extend(["--no-binary", ":all:"])
+
+        if self.python_packages:
+            logger.debug("install python packages: %s", self.python_packages)
+            for pkg in self.python_packages:
+                cmd.append(pkg)
+
+        if self.requirements:
+            logger.debug("install requirements: %s", self.requirements)
+            for req in self.requirements:
+                cmd.extend(["-r", req])
+
+        retcode = _process_run(cmd)
+        if retcode:
+            raise CommandError("problems installing dependencies")
+
+
+def _process_run(cmd, **kwargs) -> int:
+    logger.debug("Running external command %s", cmd)
+    try:
+        ret = subprocess.run(cmd, **kwargs)
```

Comment:
Shouldn't the output be captured and reported "in debug"?

Reply:
The tool is executed using `polite_exec()` in the builder command, so its output is handled by the caller.